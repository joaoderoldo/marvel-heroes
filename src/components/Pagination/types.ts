export interface PaginationProps {
  totalPages: number;
  handleOffset: (offset: number) => void;
  classes?: string | string[];
}
