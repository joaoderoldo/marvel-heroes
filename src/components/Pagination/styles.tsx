import styled from "styled-components";

export const PaginationWrapper = styled.div`
  background-color: var(--white, #fff);
  border-top: 1px solid #ebebeb;
  min-height: 60px;
  display: flex;
  align-items: center;

  .pagination-container {
    display: flex;
    justify-content: center;
    align-items: center;
  }
`;

export const PaginationItem = styled.div`
  width: 40px;
  height: 40px;
  font-weight: 500;
  font-size: 0.875rem;
  color: #707070;
  transition: background-color 0.2s ease-in-out, color 0.2s ease-in-out;
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;

  &:hover,
  &.active {
    background-color: var(--primary-dark, black);
    color: var(--white, #fff);
  }
`;
