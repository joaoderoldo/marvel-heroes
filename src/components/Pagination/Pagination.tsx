import { useState, useEffect } from "react";
import { PaginationWrapper, PaginationItem } from "./styles";
import { fetchPageNumbers } from "@/components/Pagination/utils/helperPagination";
import { Container } from "@/components";
import { useCallback } from "react";

const Pagination = ({ totalPages, handleOffset, isHidden, ...props }) => {
  const [currentPage, setCurrentPage] = useState(1);
  const pageNeighbours = Math.max(1, Math.min(0, 2));
  const [pages, setPages] = useState([]);

  useEffect(() => {
    if (!Number.isNaN(totalPages) && currentPage === 1) {
      const initialPagination = fetchPageNumbers({
        currentPage: 1,
        totalPages,
        pageNeighbours,
      });

      setPages(initialPagination);
    }
  }, [totalPages]);

  const goToPage = useCallback(
    (page: number) => {
      const current = Math.max(0, Math.min(page, totalPages));
      const paginationData = fetchPageNumbers({
        currentPage: current,
        totalPages,
        pageNeighbours,
      });
      setPages(paginationData);
      setCurrentPage(current);
      const offsetValue = (page - 1) * 7;
      handleOffset(offsetValue);
    },
    [pageNeighbours, totalPages]
  );

  if (isHidden) return null;

  return (
    <PaginationWrapper {...props}>
      <Container className="pagination-container">
        {pages.length &&
          pages.map((page) => {
            if (page === "LEFT") {
              return (
                <PaginationItem
                  onClick={() => {
                    goToPage(currentPage - 1);
                  }}
                  key={page}
                >
                  <span aria-hidden="true">&laquo;</span>
                </PaginationItem>
              );
            }

            if (page === "RIGHT") {
              return (
                <PaginationItem
                  onClick={() => {
                    goToPage(currentPage + 1);
                  }}
                  key={page}
                >
                  <span aria-hidden="true">&raquo;</span>
                </PaginationItem>
              );
            }

            return (
              <PaginationItem
                key={page}
                className={`${currentPage === page ? " active" : ""}`}
                onClick={() => {
                  goToPage(+page);
                }}
              >
                {page}
              </PaginationItem>
            );
          })}
      </Container>
    </PaginationWrapper>
  );
};

export default Pagination;
