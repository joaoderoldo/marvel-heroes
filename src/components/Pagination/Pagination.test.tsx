import { vi } from "vitest";
import { render, fireEvent, getByText, waitFor } from "@testing-library/react";
import Pagination from "./Pagination";

describe("Pagination", () => {
  const handleOffset = vi.fn();

  it("should render the component", () => {
    const { asFragment } = render(
      <Pagination totalPages={10} handleOffset={handleOffset} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render pagination items correctly", () => {
    const { getByText } = render(
      <Pagination totalPages={5} handleOffset={handleOffset} />
    );
    expect(getByText("1")).toBeInTheDocument();
    expect(getByText("2")).toBeInTheDocument();
    expect(getByText("3")).toBeInTheDocument();
    expect(getByText("4")).toBeInTheDocument();
    expect(getByText("5")).toBeInTheDocument();
  });

  it("should call handleOffset function when a pagination item is clicked", () => {
    const { getByText } = render(
      <Pagination totalPages={10} handleOffset={handleOffset} />
    );
    fireEvent.click(getByText("2"));
    expect(handleOffset).toHaveBeenCalledWith(7);
  });

  it("should disable left arrow button when the first page is active", () => {
    const { queryByText, getByText } = render(
      <Pagination totalPages={10} handleOffset={handleOffset} />
    );

    const leftArrowButton = queryByText("Â«");
    expect(leftArrowButton).not.toBeInTheDocument();
    fireEvent.click(getByText("5"));
    waitFor(() => expect(leftArrowButton).toBeInTheDocument());
  });
});
