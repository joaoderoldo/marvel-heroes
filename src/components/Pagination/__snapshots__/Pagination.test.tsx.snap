// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination > should render the component 1`] = `
<DocumentFragment>
  <div
    class="sc-iAEyYk dpWWQo"
  >
    <div
      class="sc-gueYoa gEcUkb pagination-container"
    >
      <div
        class="sc-fsQiph kqgSP  active"
      >
        1
      </div>
      <div
        class="sc-fsQiph kqgSP"
      >
        2
      </div>
      <div
        class="sc-fsQiph kqgSP"
      >
        3
      </div>
      <div
        class="sc-fsQiph kqgSP"
      >
        4
      </div>
      <div
        class="sc-fsQiph kqgSP"
      >
        5
      </div>
      <div
        class="sc-fsQiph kqgSP"
      >
        <span
          aria-hidden="true"
        >
          Â»
        </span>
      </div>
      <div
        class="sc-fsQiph kqgSP"
      >
        10
      </div>
    </div>
  </div>
</DocumentFragment>
`;
